### ISSUE TITLE
Refactor: merge EGGROLL + single-net world models into one EnsembleMLP (backprop or EGGROLL trainer)

### ISSUE THREAD

**Issue by james-harvey-git** [james-harvey-git/model-based-eggroll#37]
## Context

We currently have **three** world-model classes split along confusing lines:

| Class | File | What it is |
|---|---|---|
| `MLEEnsemble` | `world_models/mle.py` | K independent **Flax** networks, NLL + AdamW. The Unifloral baseline. |
| `MLEDynamicsNet` | `world_models/mle_dynamicsnet.py` | A **single** `DynamicsNet` (EGGROLL primitive), NLL + AdamW. Stage-1 of the #30 hybrid handoff; `predict_ensemble` returns N *copies* (no diversity). |
| `EGGROLLEnsemble` | `world_models/eggroll.py` | A **single** `DynamicsNet` trained by EGGROLL; "population-as-ensemble" at inference (perturbed members). |

Three problems:

1. **Bad names.** Every world model is trained with an MLE objective, so `mle`/`MLE...` is redundant noise. The real axis we compare is the **training algorithm: backprop vs EGGROLL.**
2. **Weird split.** `EGGROLLEnsemble` and `MLEDynamicsNet` are *the same network* (`DynamicsNet`) trained two different ways — they should be one object, not two classes.
3. **The EGGROLL ensemble was never a real ensemble.** "Population-as-ensemble" produces members that are ephemeral perturbations off a single net, not genuinely-independent models, and they can't be checkpointed as N members.

## Goal

Merge `EGGROLLEnsemble` + `MLEDynamicsNet` into **one** object — proposed name **`EnsembleMLP`** — that is:

- a **genuine ensemble of N independent `DynamicsNet` members**, each randomly initialised *differently* (exactly like Unifloral's backprop ensemble — diversity from independent init, not from perturbations);
- **trainable by either algorithm** (backprop *or* EGGROLL), selected by config;
- backed by **one shared checkpoint format**, so a checkpoint trained by *either* algorithm can be loaded for fine-tuning by *either* algorithm;
- `predict_ensemble` behaves exactly like the Unifloral baseline: returns `(num_elites, obs_dim + 1)` genuinely-diverse member predictions.

The Flax `MLEEnsemble` baseline stays functionally **untouched** (it can't be EGGROLL-trained), but is **renamed** to make its provenance clear: `UnifloralEnsembleMLP`.

> Backwards compatibility is explicitly **not** a goal. Old checkpoints/configs may break; prefer the clean design over migration shims.

---

## Proposed naming (please redline)

| Old | New |
|---|---|
| `world_models/eggroll.py::EGGROLLEnsemble` + `world_models/mle_dynamicsnet.py::MLEDynamicsNet` | `world_models/ensemble_mlp.py::EnsembleMLP` |
| `world_models/mle.py::MLEEnsemble` | `world_models/unifloral_ensemble_mlp.py::UnifloralEnsembleMLP` |
| `configs/world_model/eggroll_ensemble.yaml` + `mle_dynamicsnet.yaml` | `configs/world_model/ensemble_mlp_eggroll.yaml` + `ensemble_mlp_backprop.yaml` |
| `configs/world_model/mle.yaml` | `configs/world_model/unifloral_ensemble_mlp.yaml` |

`DynamicsNet` (the per-member architecture in `eggroll/networks.py`) keeps its name — it's the single-member net the ensemble is built from.

---

### `EnsembleMLP` — one class, two trainers

`EnsembleMLP(obs_dim, act_dim, dataset_id, cfg)` holds an **N-member stacked `DynamicsNet` pytree** (leading axis `num_ensemble` over every parameter), built by vmapping `DynamicsNet.rand_init` over `num_ensemble` independent PRNG keys (this is what gives each member a different init, mirroring Unifloral's `split_rngs={"params": True}`).

`cfg.trainer: eggroll | backprop` selects the training path; `train()` dispatches to `_train_backprop` or `_train_eggroll`. The two default configs differ only in `trainer` plus the algorithm-specific block.

Shared inference interface (identical for both trainers, matches the `EnsembleDynamics` ABC and the Unifloral baseline):

- `predict_ensemble(obs, action) -> (mean, std)` each `(num_elites, obs_dim+1)`, by vmapping the **unperturbed** member forward over the elite members. Genuine epistemic diversity.
- `step(obs, action, rng)` — pick a random elite, sample `N(mean, exp(0.5·logvar))`. (Preserves the MOPO contract: `mopo.py` reads `ensemble_mean.shape[0]` and samples an index — unchanged.)
- `compute_val_mse(dataset)` — elite-mean MSE, matching the Unifloral baseline's `val_mse_elite`.

### Backprop trainer (`_train_backprop`)

Port `MLEEnsemble._train_dynamics` from Flax onto the stacked `DynamicsNet` members:

- one `optax` optimiser over the stacked N-member params;
- per-step loss = NLL summed over the ensemble axis, mean over batch/features (matching Unifloral's `mse_loss.sum(0).mean()` so members are trained jointly but diverge via independent init);
- same minibatch for all members each step (Unifloral convention).

Solver configurable via `cfg.optimizer` / `cfg.optimizer_kwargs` (reuse `resolve_optax_solver`, as `MLEDynamicsNet` already does).

### EGGROLL trainer (`_train_eggroll`) — **independent per-member ES**

Each member runs its **own** EGGROLL search — own perturbation population, own fitness — `vmap`-ped over the `num_ensemble` members. Concretely the EGGROLL state gets a leading member axis over `params`, `opt_state`, and **`es_tree_key`** (distinct keys → independent perturbation directions per member). Shared across members (as closures / broadcast): the solver, `frozen_params`, `iterinfos`, and the `sigma`/`lr` schedules.

Each step, per member: sample its own prompt transitions → vmapped population forward → per-perturbation regularised-NLL fitness → `convert_fitnesses` → `do_updates`. Diversity comes from independent init **and** independent ES noise per member.

> **Research risk / open implementation detail.** Independent per-member ES means forward-evals scale **~N×** the current single-net cost (today `population_size=4096`; 7 members naively ⇒ ~28k evals/step). `population_size` becomes **per-member** — expect to reduce it and/or the per-member step budget. Whether each member should sample *independent* prompt batches (more diversity, recommended) vs a shared batch is a knob to settle during implementation.

### Elite selection (kept)

`num_ensemble` members trained; the best `num_elites` by validation MSE are kept for `predict_ensemble`/`step`, applied **identically by both trainers**. Unlike the Unifloral baseline (which prunes non-elites on save), `EnsembleMLP` **persists all `num_ensemble` members plus `elite_idxs`**, so fine-tuning can resume the full ensemble and re-select elites. `predict_ensemble`/`step` index by `elite_idxs`.

### Single-network case (`num_ensemble = 1`)

`num_ensemble=1` is a first-class supported configuration: it **reproduces today's single-network EGGROLL world model**. The EGGROLL trainer simply vmaps over a member axis of length 1; **training and `step()` (unperturbed base + aleatoric Gaussian) are identical to the current `EGGROLLEnsemble`.** The only behavioural difference vs today is that `predict_ensemble` no longer manufactures a population-as-ensemble pseudo-spread — with one member there is one member.

**MOPO at `num_ensemble=1`:** verified against `mopo.py:100`. The penalty is `max` over members of each member's *predicted* (aleatoric) std (`exp(0.5·logvar)`), **not** a variance-across-member-means (epistemic) penalty. At one member it stays shape-valid and well-defined (`num_elites=1`, penalty = that model's predicted std norm) — it does **not** break — but it degrades to a purely aleatoric, single-model penalty. A strong epistemic MOPO penalty requires `num_ensemble>1` (genuine members), which is what the ensemble upgrade provides.

### Fine-tuning from a pretrained checkpoint (cross-trainer)

Both trainers accept `cfg.init_checkpoint` (an `EnsembleMLP` checkpoint from *either* algorithm) + `cfg.reset_optax_state`:

- Load the N-member `params` pytree as the starting point (assert pytree structure matches the configured `DynamicsNet` — same `hidden_dims`/`activation`/`init_scheme`/`backbone`/`num_ensemble`).
- **Optimiser-state carryover** only when the solver structure matches per-member; otherwise require `reset_optax_state: true` with a clear error (generalises the existing #30 guard, now in both directions).

### Validation-set parity (kept, generalised to cross-trainer)

You chose to keep continuous val-MSE curves across pretrain→finetune. Because the checkpoint can now come from *either* trainer, replace the fragile trainer-specific split-chain replay with a **single pure helper** used by both trainers:

```
derive_train_val_split(dataset, validation_split, seed) -> (train, val)
```

The held-out val set becomes a pure function of `(dataset, validation_split, seed)` — trainer-independent. The checkpoint records `seed` + `validation_split`; a fine-tune run re-derives the *same* partition from the checkpoint's values. This localises the "DO NOT REORDER" sensitivity to one helper and makes backprop↔EGGROLL parity automatic in both directions.

### `experiments/world_model.py` cleanup

Collapse the current 3-branch `isinstance` dispatch + 3 checkpoint formats:

- give every world-model class a `checkpoint_state()` method (the renamed `UnifloralEnsembleMLP` gains one) so the runner is just `{**common, **world_model.checkpoint_state()}` with **no `isinstance` branching**;
- `updates_per_epoch` becomes trainer-aware (backprop / Unifloral baseline = `batches_per_epoch`; EGGROLL = 1) via the cfg, not class identity;
- the existing `init_checkpoint` dataset/shape guard is already generic — keep it.

### `experiments/policy.py` fix (latent bug)

`policy.py:36` hardcodes `MLEEnsemble.load_from_checkpoint(...)`, so the policy stage can't consume an EGGROLL-trained world model today. Make WM loading **dispatch on the class recorded in the checkpoint** (`world_model_cfg._target_`), supporting both `EnsembleMLP` and `UnifloralEnsembleMLP`.

---

## Files to change

| File | Change |
|---|---|
| `src/mbrl/world_models/ensemble_mlp.py` | **NEW** — `EnsembleMLP`: N independent `DynamicsNet` members, `_train_backprop` + `_train_eggroll`, elite selection, cross-trainer `load_from_checkpoint`/`checkpoint_state`. |
| `src/mbrl/world_models/eggroll.py` | **DELETE** — logic merged into `ensemble_mlp.py`. |
| `src/mbrl/world_models/mle_dynamicsnet.py` | **DELETE** — logic merged into `ensemble_mlp.py`. |
| `src/mbrl/world_models/mle.py` | **RENAME** → `unifloral_ensemble_mlp.py`; `MLEEnsemble` → `UnifloralEnsembleMLP`; add `checkpoint_state()`. Behaviour otherwise untouched. |
| `src/mbrl/world_models/base.py` | (Optional) add `checkpoint_state` to the ABC. |
| `src/mbrl/experiments/world_model.py` | Collapse dispatch via `checkpoint_state()`; trainer-aware `updates_per_epoch`; keep generic `init_checkpoint` guard. |
| `src/mbrl/experiments/policy.py` | Resolve WM class from the checkpoint (drop hardcoded `MLEEnsemble`). |
| `configs/world_model/ensemble_mlp_eggroll.yaml` | **NEW** (`trainer: eggroll`). |
| `configs/world_model/ensemble_mlp_backprop.yaml` | **NEW** (`trainer: backprop`). |
| `configs/world_model/unifloral_ensemble_mlp.yaml` | **RENAME** from `mle.yaml`; update `_target_`. |
| `configs/world_model/eggroll_ensemble.yaml`, `mle_dynamicsnet.yaml` | **DELETE**. |
| `configs/experiment/halfcheetah_medium_eggroll.yaml` | Update `world_model` reference. |
| `configs/sweep/{eggroll_wm_halfcheetah,_finetune,_random,_schedule}.yaml`, `mle_dynamicsnet_halfcheetah_sgd_lr.yaml` | Update `world_model=` overrides + param paths to new names. |
| `tests/test_world_models.py` | Rename/merge into `TestEnsembleMLP{Backprop,Eggroll}` + `TestUnifloralEnsembleMLP`; generalise `TestHybridHandoff` to cross-trainer both directions; new tests for genuine member diversity + cross-trainer checkpoint round-trip. |
| `CLAUDE.md` | Update the Project Structure file listing. |

---

## Out of scope (separate follow-up issue)

The "clean up comments to match the stored comment conventions" pass is tracked separately.

## Acceptance criteria

- [ ] One `EnsembleMLP` class trains via both `trainer: eggroll` and `trainer: backprop`.
- [ ] `predict_ensemble` returns `num_elites` genuinely-diverse members for **both** trainers (a test asserts members differ).
- [ ] A checkpoint from either trainer loads into either trainer for fine-tuning (round-trip test, both directions).
- [ ] Val-MSE parity across pretrain→finetune holds for all four trainer pairings, via the shared split helper.
- [ ] Policy stage runs against an `EnsembleMLP` (EGGROLL-trained) world model.
- [ ] `num_ensemble=1` + `trainer: eggroll` reproduces current single-net EGGROLL training/`step` behaviour; MOPO runs against it (degraded but functional).
- [ ] `MLEEnsemble`→`UnifloralEnsembleMLP` baseline numbers unchanged.
- [ ] `ruff` + `pyright` + `pytest` green.
